lib/Dialect/Pipeline/Transforms/ScheduleLinearPipeline.cpp

```diff
@@ -0,0 +1,177 @@
+//===- ScheduleLinearPipeline.cpp - Linear pipeline scheduling pass -----*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// Contains the definitions linear pipeline scheduling pass.
+//
+//===----------------------------------------------------------------------===//
+
+#include "PassDetails.h"
+#include "circt/Analysis/DependenceAnalysis.h"
+#include "circt/Analysis/SchedulingAnalysis.h"
+#include "circt/Dialect/SSP/SSPOps.h"
+#include "circt/Dialect/SSP/Utilities.h"
+#include "circt/Scheduling/Algorithms.h"
+#include "circt/Scheduling/Utilities.h"
+#include "circt/Support/BackedgeBuilder.h"
+#include "llvm/Support/Debug.h"
+
+#include <iostream>
+
+#define DEBUG_TYPE "pipeline-schedule-linear"
+
+using namespace mlir;
+using namespace circt;
+using namespace pipeline;
+
+namespace {
+
+class ScheduleLinearPipelinePass
+    : public ScheduleLinearPipelineBase<ScheduleLinearPipelinePass> {
+public:
+  void runOnOperation() override;
+};
+
+} // end anonymous namespace
+
+// Returns true if 'op' should be ignored in the scheduling problem.
+static bool ignoreOp(Operation *op) {
+  return isa<hw::ConstantOp, pipeline::ReturnOp>(op);
+}
+
+void ScheduleLinearPipelinePass::runOnOperation() {
+  auto pipeline = getOperation();
+
+  // Get operator library for the pipeline.
+  auto opLibAttr = pipeline->getAttrOfType<FlatSymbolRefAttr>("operator_lib");
+  if (!opLibAttr) {
+    pipeline.emitError("missing 'operator_lib' attribute");
+    return signalPassFailure();
+  }
+  auto opLib = dyn_cast_or_null<ssp::OperatorLibraryOp>(
+      SymbolTable::lookupNearestSymbolFrom(pipeline->getParentOp(), opLibAttr));
+  if (!opLib) {
+    pipeline.emitError("operator library '") << opLibAttr << "' not found";
+    return signalPassFailure();
+  }
+
+  auto stageOpIt = pipeline.getOps<PipelineStageOp>();
+  auto stageRegOpIt = pipeline.getOps<PipelineStageRegisterOp>();
+
+  if ((std::distance(stageOpIt.begin(), stageOpIt.end()) != 0) ||
+      (std::distance(stageRegOpIt.begin(), stageRegOpIt.end()) != 0)) {
+    pipeline.emitError("Pipeline cannot have any stages or stage registers.");
+    return signalPassFailure();
+  }
+
+  // Load operator info from attribute.
+  auto problem = Problem::get(pipeline);
+
+  DenseMap<SymbolRefAttr, Problem::OperatorType> operatorTypes;
+  SmallDenseMap<StringAttr, unsigned> oprIds;
+
+  // Set operation operator types.
+  Operation *lastOp = nullptr;
```

Comment:
Can't you just use the terminator?

Reply:
If so, we'd have to include the return op in the scheduling problem (and thus also define an operator type for it).

Reply:
Is it guaranteed that the last operation in the block transitively depends on every other op? Then, it doesn't make a difference.

Otherwise (and I agologize that I overlooked that earlier), you have to insert auxiliary deps from ops without users in the graph, to the last operation. In that situation, I think that it would be easier and cleaner to treat the pipeline's terminator as a "virtual" sink node in the dependence graph. And as this modeling is intrinsic to the pass, I'd construct its operator type ad-hoc (so not loading that from the operator lib).

Reply:
> Otherwise (and I agologize that I overlooked that earlier), you have to insert auxiliary deps from ops without users in the graph, to the last operation. In that situation, I think that it would be easier and cleaner to treat the pipeline's terminator as a "virtual" sink node in the dependence graph. And as this modeling is intrinsic to the pass, I'd construct its operator type ad-hoc (so not loading that from the operator lib).

This makes sense to me - I'll give it a shot!